Groebner extension

yizhou7 opened the pull request.

Chris-Hawblitzel committed:
Add Docs/Linear directory with documentation on the linear type system

Chris-Hawblitzel committed:
Fix some typos in Docs/Linear/TypingRules.md

Chris-Hawblitzel committed:
Minor reformatting in Docs/Linear/TypingRules.md

parno committed:
Minor fixes

Chris-Hawblitzel committed:
Add Docs/Linear/RelatedWork.md

Chris-Hawblitzel committed:
Add Rust comparison to Docs/Linear/RelatedWork.md

Chris-Hawblitzel committed:
Minor fixes to Docs/Linear/RelatedWork.md

Chris-Hawblitzel committed:
Minor formatting fixes to Docs/Linear/RelatedWork.md

Chris-Hawblitzel committed:
Minor bug fixes in Resolver/Linear

Chris-Hawblitzel committed:
Allow "linear var D() := ...;" to deconstruct empty linear datatype
cases

Chris-Hawblitzel committed:
Add Docs/Linear/Library.md

Chris-Hawblitzel committed:
Add Docs/Linear/Implementation.md

Chris-Hawblitzel committed:
Add Docs/Linear/PossibleChanges.md and Docs/Linear/README.md

parno committed:
Closes #15

parno committed:
Merge branch 'cpp' of github.com:secure-foundations/dafny into cpp

parno committed:
Merge branch 'master' into cpp

parno committed:
Remove some outdated notes to simplify comparisons between branches

parno committed:
Remove project files that are no longer needed

parno committed:
This LIMITATIONS file has been absorbed into Docs/Compilation/Cpp.md

parno committed:
Merge branch 'cpp' into inout

parno committed:
Many fixes to restore everything to a state that compiles

parno committed:
Attempt to resolve a lookup in GetTypeOfFunction

parno committed:
Attmped fix to contract assertion failure.  It looks like someone added
a special exception on the `inout` branch for the "_default" module vs.
the original master code, due to the presence of rank_is_less_than.
Rustan then made the original logic more complex in support of
"Revise export reference types", commit 3a9524f5ea629c9b35307.

parno committed:
Fix a number of failing tests by eliminating bugs and/or making
the linear pass more complete

parno committed:
Fix another case in Linear.cs

parno committed:
Restore a portion of the parser that got lost in the merge

parno committed:
Fix a problem with emitting default values.  Addresses 6 test cases

parno committed:
Fix another bad default to address 4 more test cases

parno committed:
Remove redundant use of isGhost in favor of usage

parno committed:
Import a fix from master commit 5129ec34fa6f63

utaal committed:
Fix #13, autogenerated syntax shouldn't be resolved early: dafny refinement relies on late binding

utaal committed:
Partial fix for linearity failures in tests introduced by new auto ghost machinery

Chris-Hawblitzel committed:
Fix linearity checking in VarDeclPattern after merge

Chris-Hawblitzel committed:
Small fixes so that Test/comp passes

utaal committed:
Add a README test for inout

Chris-Hawblitzel committed:
When resolving operator overloading, don't splice ResolvedExpression into AST before it's needed

parno committed:
Merge branch 'master' into inout-merge

parno committed:
Resolve compilation errors

parno committed:
Fix a parser bug

parno committed:
Remove some comments

Chris-Hawblitzel committed:
After merge, fix usages passed to arrow type in Dafny.atg

Chris-Hawblitzel committed:
Minor post-merge fixes

Chris-Hawblitzel committed:
Minor post-merge fix for linear tuples

Chris-Hawblitzel committed:
Restore some of Dafny.atg from before merge

Chris-Hawblitzel committed:
Minor updates to linear documentation

Chris-Hawblitzel committed:
Remove linear-notes directory

parno committed:
Partially through merging Dafny official into betr

parno committed:
Resolve compilation errors in the merged code

parno committed:
Resolve compilation errors in the merged code

Chris-Hawblitzel committed:
Fix some merge breakages in Resolver for linear checker

Chris-Hawblitzel committed:
Fix some merge breakages in Resolver for linear checker

Chris-Hawblitzel committed:
Fix some merge breakages in Resolver for linear checker

Chris-Hawblitzel committed:
Fix some merge breakages in Resolver for linear checker

utaal committed:
add inout regression test

utaal committed:
add inout regression test

utaal committed:
add inout regression test

Chris-Hawblitzel committed:
Fix parsing when using explicit "linear" or "ghost" in pattern cases

Chris-Hawblitzel committed:
Fix parsing when using explicit "linear" or "ghost" in pattern cases

utaal committed:
fixes #21: linear inout parameter causing crash in dafny

utaal committed:
fixes #21: linear inout parameter causing crash in dafny

parno committed:
Merge branch 'betr-merge-master' into betr

parno committed:
Merge commit master commit '3b5de791712' into rebuild-attempt (branching
off of inout)

parno committed:
Merge branch 'master' into betr

parno committed:
Modify the parser and AST to accept module's with functor applications.
Parsing succeeds on all of our tests.

parno committed:
Add some additional fields to keep track of resolved data.

parno committed:
Better naming

parno committed:
Merge branch 'betr' into betr-merge-betr

This combines our improved history branch (betr-merge)
with our standard (but bad history) branch betr.

parno committed:
Temporary work on Resolver.  Some of it needs to be rolled back.

parno committed:
With some clean up, very simple functor definition test runs without
crashing.

parno committed:
Extend dependency processing to handle imports and abstracts too,
factoring out the common steps in each.  Simple.dfy passes.

parno committed:
Fill in the Root and expand test case

JackMenandCameron committed:
Added support for editing linear datatypes with multiple constructors

JackMenandCameron committed:
Updated datatypes.dfy.expect to reflect new tests

parno committed:
Fix some bit rot in our local C++ files

parno committed:
Fix some bit rot in our local C++ files

parno committed:
Fix compilation of templated datatype methods

parno committed:
Don't emit an extra template argument for templated extern structs

parno committed:
Fix: Special case the declaration of empty tuples

parno committed:
Fix: Use compile name for get_default type parameter

parno committed:
Fix tuple lit test

parno committed:
Fix: Tuple accessor was previously failing when used on tuples of
generic types.  The fix is a truly bizarre C++-ism.

parno committed:
Special case the handling of empty tuple values

parno committed:
Fix: Don't emit type arguments for static method invocations

parno committed:
Fix: Filter typeArgs for functions as well.  Turn off typeArgs even for
non-static member invocations.

parno committed:
Fix: Include module name when accessing constants

parno committed:
Simple declaration of a functor and trivial application work (see
Simple.dfy).

parno committed:
Filled in a first approximation at Functor application in ResolveModuleQualifiedId.
Need to get a signature for each module actual passed into the FunctorApplication.

parno committed:
We have a reasonably credible functor application at this point,
and we run to completion on Simple.dfy, but we hit:
        Error: module 'Output' does not declare a type 'Key'
so more work is needed

parno committed:
Improve test and add some extra module definition resolution.
Now we've reached the point where Boogie names aren't lining up.

parno committed:
Current prefix of Simple.dfy verifies!

parno committed:
Fix a bug in ResolveModuleQualifiedId so that virtual module lookup sets
both the decl and sig.  Now doing the same functor application twice
seems to work.

parno committed:
Basic refinement argument test passes

parno committed:
Working on improving module argument checking

JackMenandCameron committed:
Merge branch 'betr' into cpp-mult-const

parno committed:
Fix positive and negative checks for passing the right kinds of module
arguments. Discovered a problem when we apply a functor with ABase
and later with a refinement of ABase

parno committed:
Merge pull request #28 from secure-foundations/cpp-mult-const

Cpp mult const

parno committed:
Rolling back visibility scope duplication works surprisingly well.
Main problem is the error message for a bad application is poor.

parno committed:
Fix what appears to be a bug in official Dafny.  Now, all of Simple's
(uncommented) positive and negative tests are passing!

parno committed:
I think all of Simple's tests pass

parno committed:
Split Simple into Good and Bad test cases.  This facilitates testing the
full pipeline in the Good cases, b/c the Bad cases bail out early.

parno committed:
Lit files

parno committed:
Add lit files

parno committed:
Fix up ToString conversion for ModuleQualifiedId and FunctorApplication

parno committed:
ModuleApplication works

parno committed:
Capture the next case to work on in SimpleGood

parno committed:
Fix ToString for non-functors

parno committed:
Still debugging the use of module formals in refines functor
applications

parno committed:
Weaken filter for ModuleQualifiedId's in refinements.
Simple* verify, now with FunctorAppParamRefiner passing.

parno committed:
Trying some variants of Resolve

parno committed:
Some experiments with Resolve2.dfy

parno committed:
Another variant gives a different error

utaal committed:
disable subsequence sharing optimisation

parno committed:
Create a small test case of deep functor application (currently fails)

parno committed:
Start fleshing out deep functor app.  Need to refactor functor app, so
that we can recurse.

parno committed:
Partway through recursive application

parno committed:
Clean up and fixes so that recursive functor application works,
at least enough for SimpleGood test

parno committed:
Confirm the rest of the known tests still work

parno committed:
Resolve2.dfy now works too!

parno committed:
ResolveBadRefine also works.

parno committed:
Add an expect file for Resolve2.dfy which is now passing

parno committed:
Some comments on what's being attempted here

parno committed:
Try to do more sophisticated refines checking, despite difficult
circumstances

parno committed:
SimpleGood example seems to verify.  Resolve3 still doesn't

parno committed:
Add an (unsuccessful) attempt at resolving B(C).D when used as a
parameter.

parno committed:
For now, we don't care about Resolve.dfy or Resolve3.dfy

parno committed:
Somewhat convoluted solution to refinements of functors.  Leads a new
problem with CrashSafeSpec

parno committed:
Add new test cases

parno committed:
Slight change in SystemTheorem signature is enough for FunctorParty to
go through.

parno committed:
Use a more up-to-date version

tjhance committed:
enable inout translations on method members of opaque types

parno committed:
Implement functor output origin tracking and improve propagation of
refines clauses.  FunctorParty now passes.

parno committed:
Created a small example of refinement type confusion

parno committed:
Deal with some of StateMachineRefinement's complaints about misnamed
parameters and resolution failures.

parno committed:
Make ComposeRefinements abstract

parno committed:
Add a hack to deduplicate translated types that may have been induced by functor instantiation

parno committed:
Fix some bugs in StateMachineRefinement.dfy.  Uncomment to the point
where we're hitting complaints about concrete and abstract modules.

parno committed:
Confirm that lit tests still pass

parno committed:
Implemented policy that output of abstract functor is abstract, and
output of concrete functor with any abstract arguments is also abstract

parno committed:
Working through abstract/concrete issues

parno committed:
Problematic example from Rob

parno committed:
Fix timeLimitMultiplier by having it generate internal annotations,
rather than overwrite the original timeLimitMultiplier annotation.
Closes #1224.

utaal committed:
Revert "disable subsequence sharing optimisation"

This reverts commit d2ede02192a1d552c317661f0cf7ac1dc2b3ceb3.

Chris-Hawblitzel committed:
Disable multi-assignment to linear variables

Chris-Hawblitzel committed:
In linear/README.dfy, restrict peek to non-empty types

parno committed:
Look for timeLimitMultiplier in inductive datatypes too, not just in
classes.

parno committed:
Manually clone the CompileSignature field, so that Rob's parser example
will go through.

parno committed:
Add lit header and expect file

parno committed:
Try declaring that Functors are abstract for compilation purposes.

parno committed:
Exploring translation problem

parno committed:
Allow callers to decide whether to preserve Functor status.  Explicitly
do not do so when doing ApplyFunctor.  Solves the basic SimpleGood test.

parno committed:
All lit tests, including full SimpleGood, are passing.

parno committed:
Resolve the issue with SimpleGood's abstract test. StateMachine still
failing

parno committed:
Modify tests

parno committed:
Track the fake AliasModuleDecls that we added, so that we only
update them once during ApplyFunctor.  This suffices to let
StateMachineRefinement pass (after fixing some bugs in
StateMachineRefinement).

parno committed:
StateMachineRefinement first half back in a working state

parno committed:
More of StateMachineRefinement passes

parno committed:
Final module in StateMachineRefinement still needs some work.

parno committed:
Another failing case from Rob's parser

parno committed:
A version of Dafny that is successful on refine.dfy and parser.dfy.
But more work is needed to generalize to possible future instances.

parno committed:
Refactor refinement updates, so we can recurse if needed.
Still need to think more about what happens when we refine
a module that includes regular module imports that are themselves
functors.  All lit tests pass.

parno committed:
Minor changes

parno committed:
Improvements to UpdateRefinement, and a refactoring
of UpdateFunctorApp so that Rob's parser example no longer
crashes.  Instead it complains about comparable types.

parno committed:
Fixes the recursion in UpdateRefinment, so that parser.dfy will pass,
assuming the last module is made abstract.

parno committed:
Treat functor formals as a form of concrete module.  A version
of parser.dfy passes

parno committed:
Always generate _Compile versions of Functor modules,
and try harder to propagate CompileSignatures.
Together, this enables our new Compile test to pass.
Previously, it failed with a mysterious type-mismatch error,
which boiled down to comparing ABase.Key with ABase_Compile.Key

parno committed:
Simplify parser example down to four modules

parno committed:
Fix the simplification

parno committed:
Create _Compile versions of everything.  With this change,
simplified parser verifies.

parno committed:
Full parser works too

parno committed:
Even fuller parser

parno committed:
Debugging SimpleGood case.

parno committed:
Refactored the code that handles _Compile modules into a separate
procedure that we can call explicitly in ApplyFunctor.  This seems to
have solved the immediate problem with SimpleGood

parno committed:
Restore rest of SimpleGood

parno committed:
One more minor extra test

parno committed:
Full StateMachineRefinement passes!

parno committed:
PCM bugs reported by Travis

parno committed:
Try disabling translation of virtual modules

parno committed:
Increase visibilityScopes
The following tests succeed but verify a lot more:
  Compile.dfy (13 vs 1)
  Resolve2.dfy (27 vs 2)
  parser.dfy (61 vs 3)
  SimpleGood (99 vs 9)
  PCM2

Still failing:
  FunctorParty
  StateMachineRefinement
  PCM1

parno committed:
Try filtering _System out of scopes, so that we don't try to verify
_System members.  We still have failing files:
    modules/FunctorParty.s.dfy
    modules/Resolve2.dfy
    modules/StateMachineRefinement.dfy
    modules/parser.dfy
Two (parser and Resolve2) are due to mismatches in the number verified.
That's because they keep reverifying things like NativeTypes.byte.

parno committed:
Revert "Try filtering _System out of scopes, so that we don't try to verify"

This reverts commit 845ecf446451e1e8dbcd9bbb7cb258025173f8be.

parno committed:
Revert "Increase visibilityScopes"

This reverts commit 06384fb9127f6b9c63823429e459e653515daf02.

parno committed:
Revert "Try disabling translation of virtual modules"

This reverts commit 1e1c85820c1049b079baada3805ffd63d4af85af.

parno committed:
Try alternate formulation of PCM1.  That works.  PCM2 still has issues.

parno committed:
Simplify PCM2 some more

parno committed:
Clone more resolution info and have the later portions of resolution
apply to virtual module defintiions too.  With this change, PCM2.dfy
passes.

parno committed:
Add PCM files to lit

parno committed:
Give Functor outputs unique names.  This doesn't seem to break existing
tests, causes the new PCM3 test to go through, and allows the original
formulation of PCM1

parno committed:
Bug report #33 from Travis

parno committed:
Small simplifications

parno committed:
Create a custom hasher and equality tester for FunctorApplication for
use with the virtualModules dictionary.  In particular, we ensure that
any arguments that are themselves Functors are applied, and we follow
AliasModuleDecl pointers to their actual targets before hashing or
checking for equality.

Closes #33

parno committed:
Fixes SimpleGood, but FunctorParty still fails

parno committed:
Refactor the .Origin field from Signature to LiteralModuleDecl.
This allows us to associate origins with the environment that has
the relevant definitions for names in the origin's functor app.
All unit tests now pass.

parno committed:
Add git issue 33 to lit

parno committed:
Add an extra check for module parameters that are actually functors
missing their arguments.

Closes #31.

parno committed:
Partial fix for the problem identified in Git issue #34.
Full solution would recursively explore a full alias chain.

parno committed:
When calling ApplyFunctor, always attempt to update any aliases that
involve functors, and when updating a functor, check whether the
arguments are themselves functors and update them as well.

Closes #35

parno committed:
Remove some tests we decided not to support

parno committed:
Remove another unused test

parno committed:
Bug with abstract

parno committed:
Add another condition where we won't worry about abstract modules.
Closes #36

parno committed:
New issues.

parno committed:
Rename

parno committed:
Fix complaints about use of int

parno committed:
With some bug fixes, our first compilation test succeeds.

parno committed:
Try to treat functor arguments as concrete too

parno committed:
Test case from git issue #37

parno committed:
Add a missing .expect file

parno committed:
Revamp handling of module formal and actuals.  We now explicitly keep
track of whether an actual argument in a functor application corresponds
to a formal argument of the enclosing functor, and we also track which
aliases are fake via a new Formal member.  As a result, we can move away
from the janky use of name strings for matching formals and actuals.

parno committed:
Example where more module info would be useful.

parno committed:
Print more details about resolution errors when using /titrace

parno committed:
Add a lit test for issue 32

parno committed:
When updating a functor, follow all alias decls, updating functors along
the way if needed.  Closes #40.

parno committed:
Don't assume oldActual is a ModuleActualDecl

tjhance committed:
only generate operator== and hash for types that support equality

also, attempt to fix issue where Dafny didn't correctly compute that a datatype didn't have equality support

parno committed:
Don't emit parentheses in module names.  Closes #43

parno committed:
Change directories

parno committed:
Turn off IsToBeCompiled for functor applications that make use of
ModuleActualFormals.  Closes #44.

parno committed:
Declare shared formals as C++ references.

tjhance committed:
Revert "only generate operator== and hash for types that support equality"

This reverts commit 3af85d21f091db4dc96481ea9367278fe0f5dc05.

There are enough bugs in Dafny's computation of equality-support, so this fix
causes too many problems right now. For now, workaround by defining operator==
for extern types

tjhance committed:
add get_default for int types that were missing

parno committed:
Merge branch 'mod-params-bryan' into betr

parno committed:
Restore the ability to import an export set

parno committed:
Compute function fuel over virtual modules as well

parno committed:
Test case for opaque

parno committed:
Alternate approach to generating compile names.  Removes unique counter
when naming functor applications.  Should be safe, in theory.
Closes #49

parno committed:
Don't compile modules that have arguments that are not going to be
compiled.  Closes #51

parno committed:
Part way through revamp of FunctorApplication, so we can use the fully
resolved argument names when naming the functor's output

parno committed:
Consolidate the checking of whether something should be compiled,
and use it in both the Compiler, and when the Resolver checks for
duplicate compile names.  Closes #52.

yizhou7 committed:
parser changes for gbassert

yizhou7 committed:
a skeleton for SingularDriver in Translator

yizhou7 committed:
a very basic version

yizhou7 committed:
open a subprocess for Singular

yizhou7 committed:
cannot get WaitForExit with timeout working for whatever reason

yizhou7 committed:
hardcode timeout in process start

yizhou7 committed:
handle certain sub expressions as opaque

yizhou7 committed:
use separate files for queries

yizhou7 committed:
add some use cases

parno commented on Source/Dafny/Translator.cs:
```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
+    }
+
+    string OpacifySubExpr(Expression expr) {
+      if (expr is SeqSelectExpr) {
```
C# will let you write:
```
if (expr is SeqSelectExpr seqSelExpr) {
  ...
```
i.e., you can combine the test with the naming of the new variable.  The name scoping is annoying though, each new variable needs to be unique, so you can't just call them all `e`, for example.

yizhou7 replied:
Ah that's good to know. Thanks for the tip!

parno commented on Source/Dafny/Translator.cs:
```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
+        }
+        var seq = this.OpacifySubExpr(seqSelExpr.Seq);
+        var index = this.OpacifySubExpr(seqSelExpr.E0);
+        return String.Format("{0}[{1}]", seq, index);
```
It seems like this is going to emit something like `x[5]` to Singular.  Won't that confuse it?

parno replied:
Oh wait, I see.  The goal is just to create a unique string 'key' to use later in the `opaqueExprs` dictionary?

yizhou7 replied:
Ah yes, exactly. We create a temp variable in place of the expression, and when the exact same expression shows up again, we will just use the temp variable. Singular will have next to no understanding of the expression, only knowing that it is some integer.

parno commented on Source/Dafny/Translator.cs:
```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
+        var mod = this.EncodeSubExpr(appplyExpr.Args[2]);
+        // add the mod to the generators
+        polys.Add(mod); 
+        return this.EncodeEqualityPoly(appplyExpr.Args[0], appplyExpr.Args[1]);
```
Why aren't `Args[0]` and `Args[1]` encoded here?  They were encoded in `EncodeGeneratorPoly`.

yizhou7 replied:
Oh in this case we are encoding the "question": is it true that `Args[0]` and `Args[1]` are congruent modulo `Args[2]`? So `Args[2]` goes into the ideal, while `Args[0] - Args[1]` becomes the query.

parno replied:
Right, but I was expecting something like:
```
return this.EncodeEqualityPoly(this.EncodeSubExpr(appplyExpr.Args[0]),this.EncodeSubExpr(appplyExpr.Args[1]));
```
```

yizhou7 replied:
Oh in this case the `Args[2]` needs to go into the ideal. The rest of the polys in the generators are effectively all 0, except this one. Then if the difference `Args[0] - Args[1]` can be written in terms of a linear combination of the generators, it must be a multiple of `Args[2]`, proving the congruence.

parno commented on Source/Dafny/Translator.cs:
```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
+      }
+    }
+
+    string EncodeQueryPoly(Expression expr) {
```
This looks very similar to `EncodeGeneratorPoly`.  Would it make sense to factor out the common bits?

yizhou7 replied:
That is true, I will try extract some common bits.

parno commented on Source/Dafny/Translator.cs:
```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
+        } else if (ss is AssumeStmt) {
+          this.EncodeGeneratorPoly((ss as AssumeStmt).Expr);
+        } else {
+          Console.WriteLine("Error: Unexpected Grobner poorf body statement {0}", ss);
```
"proof body"

yizhou7 replied:
Good catch. I can't English  :)

parno commented on Source/Dafny/Translator.cs:
```diff
@@ -10321,7 +10546,16 @@ void TrStmt(Statement stmt, BoogieStmtListBuilder builder, List<Variable> locals
+              var assumeFalse = new AssumeStmt(stmt.Tok, stmt.EndTok, falseBody, null);
+              assertStmt.Proof.Body.Add(assumeFalse);
+            }
+            TrStmt(assertStmt.Proof, proofBuilder, locals, etran);
```
Just to confirm, this is a Dafny level change that adds a Dafny-level "assume false" to the proof body that's supplied in the `by { ... }` clause of the assert?  And it's added at the end of the body, so that Dafny will check the assertions the developer wrote, since Singular trusts that they are correct?

yizhou7 replied:
Yes, that is exactly my intention.

parno replied:
It might be worth adding a small comment explaining the overall intent, or even showing a "before" and "after" version of some sample Dafny code.

yizhou7 replied:
Good point, will do.

parno reviewed (approved):
Do all of those examples go through?  If so, that's really cool!  I'm also impressed you could whip this together so quickly!

yizhou7 committed:
address some comments from @parno

yizhou7 committed:
fix missing variable

yizhou7 committed:
handle more opaque expressions

yizhou7 committed:
add more support for opaque operations (binop)

yizhou7 committed:
emit null as empty string in opaque

yizhou7 committed:
bug fix for OpacifySubExpr (args for function calls were not encoded correctly)

yizhou7 committed:
add groebner keyword for lemma call

yizhou7 committed:
retain support for gbassert

zhenkuny committed:
[groebner-extension]: close query file after writing

close the `queryFile` for Singular solver. Otherwise there will be IOException error when deleting it on Windows.

yizhou7 committed:
Merge pull request #54 from zhenkuny/patch-1

[groebner-extension]: close query file after writing

yizhou7 committed:
Merge branch 'betr' into groebner-extension

yizhou7 committed:
fix type errors

yizhou7 committed:
fix seq index access in gbassert
